### PR TITLE
Multi arch build for debezium (DBZ-5213)

### PR DESCRIPTION
**Pull request by nilshartmann**
This is my draft how to build multi arch (amd64 and arm64) builds for debezium (https://issues.redhat.com/browse/DBZ-5213).

Note that I do not have much experience with docker builds (esp. not with multi arch builds), so probably there are better or other ways to implement the build. But maybe my changes can be a starting point.

Changes on this branch:

- all images are built for amd64 and arm64
- all images hopefully tagged with the same semantics as before 🙏
- all images are pushed to docker hub and quay.io.
  - That behaviour differs from the current version. I think with buildx you HAVE to push the image somewhere as long as you build for more than one architecture.
  - The script could be changed, so that it would be possible to not push the image if only one architecture is specified (for local tests for example)
- I made the name of the "debezium" docker hub user configurable
  - it defaults to `debezium`
  - I don't know if that is really neccessary, but I didn't found another way to actually test the push and pull of the images. So changing the name is for testing only. If there is a better way for testing, please let me know.
- I made the name of the "second" registry (quay.io) configurable
  - it defaults to `quay.io`
  - this change is also related to testing: while testing you can specify another second registry (for example a local docker registry)
- the GH workflow actions install buildx
  - for some reason the build on github still fails, maybe someone with buildx and/or GH actions know-how can help.
  - all changes work on my local machine 🙄

### CODE REVIEW

**Comment by jcechace** [comment]
@nilshartmann  I will ask for a bit of a patience here, I will review this within next week or so (I'm currently on PTO).

@jpechane  does it make sense to retrospectively change the docker files for previous releases? Although this is just parametrisation I wouldn't do that.

**Comment by gunnarmorling** [comment]
> does it make sense to retrospectively change the docker files for previous releases?

Doing those changes for 1.9 and 2.0 definitely is enough. We don't "support" older versions anymore at this point.

**Comment by nilshartmann** [comment]
> Doing those changes for 1.9 and 2.0 definitely is enough. We don't "support" older versions anymore at this point.

What about duplicate the current scripts (`build-multi-arch-debezium.sh` for ex) and let them run with parameterized Dockerfiles only for 1.9 and upcoming releases. The current scripts then can be limited to pre-1.9 releases with current, non parameterized Dockerfiles? So changes would affect only Dockerfiles 1.9+ (and new build scripts).

**Comment by jcechace** [comment]
@nilshartmann The build-all script should likely be extended to allow specifying which components to build. Ideally it should be able to determine that on its own and skip those were ARM build is not possible. Currently I can see 2 reason why the build in GH is failing

1) THe GH action may not setup a buildx builder (not sure about this one)
2) It will for sure fail with example-mongo as mongo:3.2 doesn't have ARM manifest. The build ends with 

```bash
 => ERROR [linux/arm64 internal] load metadata for docker.io/library/mongo:3.2
...
ERROR: failed to solve: mongo:3.2: no match for platform in manifest sha256:0463a91d8eff189747348c154507afc7aba045baa40e8d58d8a4c798e71001f3: not found
```

Further attempt to explicitly pull mongo on AMR leads to cleaner error message

```bash
$docker pull mongo:3.2
3.2: Pulling from library/mongo
no matching manifest for linux/arm64/v8 in the manifest list entries
```

File: build-debezium.sh
Comment (jcechace) [inline build-debezium.sh:19]: @nilshartmann  Am I missing something? This does not look to be used anywhere else in the context of debezium build.  For testing / dry run it might be useful to specify to use a local registry as well (buildx currently doesn't support builds from local storage for multiarch build and it needs the image to be pushed). 


This also leads me to a opinion that the semantics of `DEBEZIUM_DOCKER_NAME` and `DEBEZIUM_QUAY_IO` is not clear.

**Comment by jcechace** [comment]
@nilshartmann  I might be missing few steps so if you could provide steps how to run the build (ideally with the use of `registry:2` container as a registry, it would be greatly appreciated.

File: build-debezium.sh
Comment (jcechace) [inline build-debezium.sh:77]: this set of quotes causes an issue where each `-t` element is actually passed as `'-t debezium/base:latest'` which is an incorrect expansion causing docker to complain about invalid reference format.

**Comment by nilshartmann** [comment]
Hi @jcechace! 

Thanks a lot for your feedback and comments!

 > The build-all script should likely be extended to allow specifying which components to build. Ideally it should be able to determine that on its own and skip those were ARM build is not possible (however this might not be feasible).

I will change the script. Probably first by manually setting a list of platforms for each image (or list of images that should not be built for ARM), as it seems to be much easier to implement. If anything else works, we can try to automatically determine for what images ARM should be skipped imho.

> 
> Currently I can see 2 reasons why the build in GH is failing
> 
>     1. The GH action may not setup a buildx builder (not sure about this one)

I *think* buildx build is set up, as I tried with a very simple image and that worked. But I'll re-check.

> 
>     2. It will for sure fail with example-mongo as mongo:3.2 doesn't have ARM manifest. The build ends with

Thanks! Good point!

Regarding your other questions. I introduced the env variables to be able to set different image names/registries during my local test builds:

* `DEBEZIUM_DOCKER_NAME`: I set this to my dockerhub username to test build and deploy to docker hub
* `DEBEZIUM_QUAY_IO`: While tests I set this to a local docker registry

If there is a better way to run, test and push the build/images (locally), please let me know. I'm in no way a docker expert 🥺

File: build-debezium.sh
Comment (nilshartmann) [reply build-debezium.sh:77]: Will fix that, thanks!

**Comment by jerrinot** [comment]
hello @jcechace, a long time no see! :)
I just stumbled upon this as I'm using the `debezium/connect-base` container image in my tests and I realized it's too slow because it's running emulated on M1 mac. It would be so cool to have multi-arch images. 

many thanks to @nilshartmann for this work!

**Comment by nilshartmann** [comment]
Hi @jcechace,

I pushed some new commits. Inside the repository there is a new `README-build.md` file that describes the steps for building images locally for test. 

The `build-all` script now runs:
- `build-postgres-multiplatform.sh` and `build-postgres-multiplatform.sh` for either with `linux/amd64` or `linux/amd64,linux/arm64`. I tested all versions on all platforms. If the arm build for one version failed, I deceided to stay with `linux/amd64` only. 
- Debezium Builds in `build-all` run the "old" `build-debezium.sh` for all Debezium versions that are not explizitly listed in `build-all.sh` (see there). Newer versions are built using `build-debezium-multiplatform.sh` for `linux/amd64,linux/arm64`.
(btw I had trouble building some of the older versions of Debezium, eventhough I didn't change the script. Might be a problem on Mac M1.)

You can use both env variables `DEBEZIUM_DOCKER_NAME`and `DEBEZIUM_QUAY_IO` to change the registries the images are pushed into (please see the `README_build.md` file for more information including setup of a local registry:2). 


(Maybe `DEBEZIUM_DOCKER_NAME`and `DEBEZIUM_QUAY_IO` should be renamed to something like `DEBEZIUM_DOCKER_REGISTRY_1_NAME`and `DEBEZIUM_DOCKER_REGISTRY_2_NAME`)

BTW you can build the images and push them to a local registry inside a GitPod workspace if you like. Just open the branch in GitPod, then run on terminal:
```bash
docker-compose -f local-registry/docker-compose.yml up -d

./setup-local-builder.sh 

export DEBEZIUM_DOCKER_NAME=localhost:5500/debezium
export DEBEZIUM_QUAY_IO=localhost:5500/debeziumquay

./build-postgres-multiplatform.sh 14-alpine "linux/amd64,linux/arm64"
./build-debezium-multiplatform.sh 1.9
```


Propably there is still some work to do, please let me know, if I can help. 

Thanks,
Nils

**Comment by jcechace** [comment]
> > Doing those changes for 1.9 and 2.0 definitely is enough. We don't "support" older versions anymore at this point.
> 
> What about duplicate the current scripts (`build-multi-arch-debezium.sh` for ex) and let them run with parameterized Dockerfiles only for 1.9 and upcoming releases. The current scripts then can be limited to pre-1.9 releases with current, non parameterized Dockerfiles? So changes would affect only Dockerfiles 1.9+ (and new build scripts).

Sounds like a good solution to me -- we keep the ability to build the old (but unsupported) and also add the new options. It also seems to be in order with how we duplicate dockerfiles for new versions of DBZ.

**Comment by jcechace** [comment]
@nilshartmann  Regarding the naming such as DEBEZIUM_DOCKER_REGISTRY_1_NAME... I think this would be better. Maybe even DEBEZIUM_DOCKER_REGISTRY_PRIMARY_NAME.

**Comment by nilshartmann** [comment]
Hi @jcechace, 

> Regarding the naming such as DEBEZIUM_DOCKER_REGISTRY_1_NAME... I think this would be better. Maybe even DEBEZIUM_DOCKER_REGISTRY_PRIMARY_NAME.

I changed the names to `DEBEZIUM_DOCKER_REGISTRY_PRIMARY_NAME` and `DEBEZIUM_DOCKER_REGISTRY_SECONDARY_NAME`

**Comment by jcechace** [comment]
@jerrinot 

> > although It look like it is quite slow
> 
> I noticed that too, not sure if this is somehow related to the emulator.
> 
> I wonder if it really makes sense to rebuild all images everytime (`build-all.sh`).
Technically yes, since there is no need to run the script unless there were some changes. The majority of slowness (apart from UI) is PG  builds

> > At the moment this is all a black magic for me, so I will poke around a bit before merging.
> 
> Please let me know if I can help.

I think I got a better grasp. Seems an alternative would be a cross build with arch specific builder stages, however that would require a modification of all dockerfiles so our current aproach seems more feasible for now.

**Comment by jcechace** [comment]
@nilshartmann It looks like PG 14 does not build via buildx on amd64 (it build fine using standard docker build). Another thing which seems to be failing consistently is the UI -- however I think that might be a known issue (the log output for buildx is not as informative as a regular build so I will look more into this).

**Comment by nilshartmann** [comment]
@jcechace I tried building PG 14 both on GitPod and Mac M1  using `./build-postgres-multiplatform.sh 14 linux/amd64` and it worked successfuly. Only building for `linux64/arm` failed.

**Comment by jcechace** [comment]
@nilshartmann  can you make sure that the buildx cache is pruned? Strangely enough when I built PG 14 using the old ``build-postgres.sh`` script with regular docker build, then consequent run of the multiplatform script passed. Otherwise I'm gettting 

```
#0 49.03 make: *** [/usr/lib/postgresql/14/lib/pgxs/src/makefiles/../../src/Makefile.shlib:293: decoderbufs.so] Segmentation fault
```

This happens on EC2 arm64 VM  running Ubuntu.

**Comment by nilshartmann** [comment]
@jcechace I did `docker buildx prune --all`(Mac M1) and the build still works. I also setup a test build on Github (ubuntu) and it works: https://github.com/nilshartmann/pg-buildx/actions/runs/3135577830/jobs/5091422539

**Comment by jpechane** [comment]
One remark, can we have the x86 build only by default and multiarch executed only upon an option/env var setting?

**Comment by nilshartmann** [comment]
@jpechane So when running `build-all.sh` the "old" `build-debezium`, `build-postgres` and `build-mongo` scripts should be run? Or would it be sufficient to run be default the new scripts but set the platform to `linux/amd64` only? In any case I think it would be easier to provide to scripts `build-all.sh` and `build-all-multiplatform.sh`.

**Comment by jpechane** [comment]
@nilshartmann That's an implementation detail :-). Yes, setting platform to `linux/amd64` is enough. And I agredd with having too scripts, `build-all.sh` woud produce the same output as now and `build-all-multiplatform.sh` will do evertyhing.

**Comment by jcechace** [comment]
Great idea, I think this will provide the largest amount of backward compatibility. 

Regarding the build of PG 14... it is strange, at this point I have about 60% success rate on building that particular image with buildx on arm64 EC2 machine. I wonder if something is different on Apple silicon. To make things weirder, I can see two errors consistently... 

The one mentioned above and 

```
#0 267.1 Setting up software-properties-common (0.96.20.2-2.1) ...
#0 270.4 Traceback (most recent call last):
#0 270.4   File "/usr/bin/py3compile", line 319, in <module>
#0 270.4     main()
#0 270.4   File "/usr/bin/py3compile", line 298, in main
#0 270.4     compile(files, versions,
#0 270.4   File "/usr/bin/py3compile", line 185, in compile
#0 270.4     cfn = interpreter.cache_file(fn, version)
#0 270.4   File "/usr/share/python3/debpython/interpreter.py", line 212, in cache_file
#0 270.4     (fname[:-3], self.magic_tag(version), last_char))
#0 270.4   File "/usr/share/python3/debpython/interpreter.py", line 246, in magic_tag
#0 270.5     return self._execute('import imp; print(imp.get_tag())', version)
#0 270.5   File "/usr/share/python3/debpython/interpreter.py", line 359, in _execute
#0 270.5     raise Exception('{} failed with status code {}'.format(command, output['returncode']))
#0 270.5 Exception: python3.9 -c 'import imp; print(imp.get_tag())' failed with status code 139
#0 270.6 dpkg: error processing package software-properties-common (--configure):
#0 270.6  installed software-properties-common package post-installation script subprocess returned error exit status 1
```
Nevertheless, I am also going to try it the other way around and build the packages on amd64. The issue might be with qemu interpreters on arm64 ubuntu.

**Comment by jcechace** [comment]
PG14 (non alpine) builds just fine for both amd64 and arm64 on amd64 ubuntu machine in EC2.  I will do one final pass of build-all for both platforms on amd64 host and once @nilshartmann  implements the changes @jpechane  suggest we should be good to merge this.

**Comment by nilshartmann** [comment]
With latest (force) push, I changed the scripts as discussed:

- `build-*.sh` are the "old" scripts, mostly unchanged (see below)
- `build-*-multiplatform.sh` contain the new scripts

There is one exception: `build-debezium.sh` also sets the `DEBEZIUM_DOCKER_REGISTRY_PRIMARY_NAME` 
and  `DEBEZIUM_DOCKER_REGISTRY_SECONDARY_NAME` env because those are referenced in
the 1.9 and 2.0 Dockerfiles (for kafka, connect etc.).

If you run `build-debezium.sh` for an older version than 1.9 nothing should change
comparing to the current scripts. Running `build-debezium.sh` for 1.9+ also should
not change something, but (base) image names in Dockerfiles are taken from environment
variables that are either set on commandline or set to default values inside `build-debezium.sh`

I also changed `build-debezium-multiplatform.sh` to align it with the other
 `build-*-multiplatform` scripts. It now takes the platforms to built as second
 argument. So you could limit the build for example to only `linux/amd64`.
 
Then there is `build-tools.sh` and `build-tools-multiplatform.sh`. Not sure if multiplatform support
is needed here, if not, we could remove `build-tools-multiplatform.sh` otherweise add it to GH actions.

**Comment by nilshartmann** [comment]
@jcechace

> ```
> ERROR: failed to solve: process "/dev/.buildkit_qemu_emulator /bin/sh -c echo \"$SHA256HASH /tmp/zookeeper.tar.gz\" | sha512sum -c - &&    tar -xzf /tmp/zookeeper.tar.gz -C $ZK_HOME --strip-components 1 &&    rm -f /tmp/zookeeper.tar.gz" did not complete successfully: exit code: 2
> ```

> multiarch/qemu-user-static didn't seem to work for me (ubuntu arm64 instance in EC2) while the tonistiigi/binfmt one worked. At the moment this is all a black magic for me, so I will poke around a bit before merging. So far the build is running (although It look like it is quite slow -- however it looks like the main bottleneck is downloading maven dependencies so probably unrelated to this PR).

There is another GH action (`docker/setup-qemu-action@v2`) that sets up qemu and I added that to my test build, and it seems to work. See here an example (zookeeper only, but I'm optimistic, that kafka also builds): https://github.com/nilshartmann/docker-arch-build-test/actions/runs/3138713626/jobs/5098350092

Another option would be to replace `tar -xzv` with `gunzip tarfile.tar.gz && tar -xf tarfile.tar`. I tried that for zookeeper also and it works (https://github.com/nilshartmann/docker-arch-build-test/actions/runs/3138581786/jobs/5098069252). We would have to change the Dockerfiles for apache and zookeeper (1.9+) and connect (snapshot). 

Advantage of changing the docker file over the qemu is imho, that anyone can simply build the images (locally) without having to install qemu.

**Comment by nilshartmann** [comment]
Finally I was able to build all 1.9 and 2.0 images (with the exception of debezium-ui 2.0) 😊

Downside: it took more than four hours 😱 https://github.com/nilshartmann/docker-arch-build-test/actions/runs/3138917357/jobs/5098778184

I haven't had a deeper look into it yet, but building the `tools` images took more than three hours. If anyone has an idea how to speed that up, please let me know. Otherwise I would suggest that we stay with the old build here and do not provide multiplatform images.

**Comment by jcechace** [comment]
@nilshartmann nice work, everything looks in order -- I was able to successfully build all the images as well (doing so on x86 is actually much faster (about 3 times) as emulating arm on x86 is faster). I will do some testing with produced images and merge the PR afterwards.

**Comment by nilshartmann** [comment]
I fixed the shellcheck error (https://github.com/debezium/container-images/actions/runs/3141452521/jobs/5128785789#step:3:7) with my latest push

**Comment by jcechace** [comment]
@jpechane  LGTM, one this is that the  GH Action for anonymous build will likely take quite a lot of time. Do we perhaps want to make it optional or something?  

Otherwise once the action is finished this can be merge.

**Comment by nilshartmann** [comment]
@jpechane @jcechace I wonder it it would be possible to use a matrix build in GH to build mongo, postgres and debezium (for example) in parallel. 
For this we could add an agument to `build-all-multiplatform.sh` that specifies what "artifact" to build (it did that for my own testing purposes already here: https://github.com/nilshartmann/docker-arch-build-test/blob/main/build-all-multiplatform.sh). Using the GH action matrix definition it could be possible to spawn an own build step for each of those three "artifacts". Could also be done in another pull request if needed.

**Comment by jcechace** [comment]
> @jpechane @jcechace I wonder it it would be possible to use a matrix build in GH to build mongo, postgres and debezium (for example) in parallel.
> For this we could add an agument to `build-all-multiplatform.sh` that specifies what "artifact" to build (it did that for my own testing purposes already here: https://github.com/nilshartmann/docker-arch-build-test/blob/main/build-all-multiplatform.sh). Using the GH action matrix definition it could be possible to spawn an own build step for each of those three "artifacts". Could also be done in another pull request if needed.

@jpechane  @Naros  any objections? This sounds useful and as stated, we definitively need to do something about the runtime of this action.

@nilshartmann would be nice to keep the ability to build everything, so I would like to see an option to pass in multiple "component" (with no component acting as "all").

**Comment by nilshartmann** [comment]
> > @jpechane @jcechace I wonder it it would be possible to use a matrix build in GH to build mongo, postgres and debezium (for example) in parallel.
> > For this we could add an agument to `build-all-multiplatform.sh` that specifies what "artifact" to build (it did that for my own testing purposes already here: https://github.com/nilshartmann/docker-arch-build-test/blob/main/build-all-multiplatform.sh). Using the GH action matrix definition it could be possible to spawn an own build step for each of those three "artifacts". Could also be done in another pull request if needed.
> 
> @jpechane @Naros any objections? This sounds useful and as stated, we definitively need to do something about the runtime of this action.
> 
> @nilshartmann would be nice to keep the ability to build everything, so I would like to see an option to pass in multiple "component" (with no component acting as "all").

@jcechace good idea, I will implement that.

**Comment by nilshartmann** [comment]
@jcechace with newest push, `image-builds-unauth.yml` runs four jobs in parallel: `mongo`, `postgres`, `debezium` and `tools`.

**Comment by jcechace** [comment]
Failure is cause by a known UI issue.

**Review by jcechace — APPROVED** [review]

**Comment by nilshartmann** [comment]
@jcechace  anything I can do to help this PR get merged?

**Comment by jcechace** [comment]
@nilshartmann I'm waiting for green light from @jpechane  that we can merge this or whether he wants to wait until the UI issue is resolved. From my point of view the work on this PR is done :). Thank you again.

**Comment by nilshartmann** [comment]
Hi @jcechace. I wonder if we should build the images in  `image-builds-auth.yml` also in parallel as we do in the "unauth" build? Maybe with `fail-fast` here to stop the pipeline if any image fails?

File: README-build.md
Comment (nilshartmann) [reply README-build.md:23]: fixed and added the link

**Comment by jcechace** [comment]
@nilshartmann Indeed. That action is used to build images after PR is merged so I guess we don't wont that to run for hours as well. Can you make the changes please?

**Comment by jcechace** [comment]
@nilshartmann after talking with @jpechane we decided that that``image-builds-auth.yml`` should produce just x86 images (that one is used to run after PR merge to produce new images ASAP). We will be producing arms anyway as part of nightlies in jenkins.  Could you make the changes please?

**Comment by nilshartmann** [comment]
> @nilshartmann after talking with @jpechane we decided that that`image-builds-auth.yml` should produce just x86 images (that one is used to run after PR merge to produce new images ASAP). We will be producing arms anyway as part of nightlies in jenkins. Could you make the changes please?

@jcechace Done. I reverted `image-builds-auth.yml` to the original version to use `build-all.sh`.

**Comment by jcechace** [comment]
@nilshartmann I just realised that  I'd previously missed the tooling image in my testing and unfortunately I discovered an issue in the dockerfile - it hardcodes kcctl to x86 which obviously doesn't work on arm. Unfortuantely right now there are no published binaries for arm. 

@gunnarmorling  can we release kcctl also fro arm? :) 

@jpechane I suggest that we build the tooling only for x86 for now in the workflow and extend it later once we can produce a multiplatform image 


@nilshartmann  Ca we ensure that all the multiplatform scripts have the ability to take the platform as an argument (either via script parameter or as a environment varibale)? This way we can simply specify just x86 for the tooling and easily change it later.

**Comment by nilshartmann** [comment]
@jcechace

> Ca we ensure that all the multiplatform scripts have the ability to take the platform as an argument (either via script parameter or as a environment varibale)? This way we can simply specify just x86 for the tooling and easily change it later.

I think setting the platform as an arg would only make sense for the tooling script, as in the other script `build-tool-images-multiplatform.s` the platform is already set depending on the version of postgres, mongo or debezium.

Having said that I think the easiest would be to simply run the old `build-tool-images.sh` until we have multi platform build for tools?

**Comment by nilshartmann** [comment]
@jcechace @jpechane I rebased this branch on top of the newest `main` branch. While doing that, I realized that  having both `build-all`and `build-all-multiplatform` scripts, you have to set the versions of Mongo and Postgres twice. 

To avoid that, I added the env variable `MULTIPLATFORM_PLATFORMS` to `build-all-multiplatform.sh`. You can set this variable to define what platforms should be used for multiplatform build. By setting it to `linux/amd64` all images are built one for that single platform, even if they are in the list of `XX_MULTIPLATFORM_VERSIONS`. 

Imho it now should be possible to run `build-all-multiplatform` in `image-builds-auth.yml` and setting `MULTIPLATFORM_PLATFORMS` to `linux/amd64` in order to only build x86 images there and get rid of `build-all.sh`.

**Comment by jcechace** [comment]
Everything but UI (most likely only on ARM) seems in order, so I am looking into where the issue is there. 

Once I found the issue this can be merged together with https://github.com/debezium/debezium/pull/3950

**Comment by jcechace** [comment]
@nilshartmann  the issue should be fixed with latest Debezium and UI release. Can you please rebase this once more?

**Comment by nilshartmann** [comment]
> @nilshartmann the issue should be fixed with latest Debezium and UI release. Can you please rebase this once more?

@jcechace  done!

**Comment by jcechace** [comment]
@nilshartmann  one last thing. Can we add the ability for components in ``build-debezium-mltiplatform.sh`` to overwrite default the ``$PLATFORM`` when calling ``build_docker_image`` function, whcih was set for the entire script? Once that's in also set UI for just linux/amd64 at the moment. 

The reason for this is that there are issues with bulding some npm dependencies on arm64 UBI8 (builds fine on ubuntu once make is installed -- for the ubi 8 arm image that wasn't enought though). The error is


```
#6 913.8 [INFO] npm ERR! make: cc: No such file or directory
#6 913.8 [INFO] npm ERR! make: *** [../../node-addon-api/nothing.target.mk:107: Release/obj.target/nothing/../../node-addon-api/nothing.o] Error 127
#6 913.9 [INFO] npm ERR! gyp ERR! build error
#6 913.9 [INFO] npm ERR! gyp ERR! stack Error: `make` failed with exit code: 2
#6 913.9 [INFO] npm ERR! gyp ERR! stack     at ChildProcess.onExit (/sources/ui/node_modules/node-gyp/lib/build.js:201:23)
#6 913.9 [INFO] npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:390:28)
#6 913.9 [INFO] npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12)
#6 913.9 [INFO] npm ERR! gyp ERR! System Linux 5.15.0-1020-aws
```

**Comment by jcechace** [comment]
@rk3rn3r Could somebody of the UI guys look into building the UI inside UBI8 arm based container? I think just building the container from current Dockerfile on arm machine

**Comment by nilshartmann** [comment]
Hi @jcechace,

with latest push you can now set a platform for each component in `build-debezium-multiplatform` individualy:

```
export DEBEZIUM_UI_PLATFORM=linux/amd64
```

Naming convention for platform env variable: component name in uppercase, replace `-` with `_`.

If no variable is set for a component the default platform passed to the script is used (as before).

Hope that helps!

**Comment by jcechace** [comment]
@nilshartmann Thanks, although I managed to debug the issue with 2.0 UI build, this will still be useful. Can you please also add ``make gcc-c++`` to line 17 in ``ui/2.0/Dockerfile`` in order to keep this in single PR? Seems npm install needs to do some compiling on arm and these tools are not part of minimal UBI9 image base. 

.. and apology for dragging this for so long, based on my local test runs this should truly be the last thing in order to get everything working :)

**Comment by nilshartmann** [comment]
> @nilshartmann Thanks, although I managed to debug the issue with 2.0 UI build, this will still be useful. Can you please also add `make gcc-c++` to line 17 in `ui/2.0/Dockerfile` in order to keep this in single PR? Seems npm install needs to do some compiling on arm and these tools are not part of minimal UBI9 image base.

@jcechace done!


> .. and apology for dragging this for so long, 

No problem! When opening this PR i didn't expected this to be a quick+easy thing. Thanks for your support and testing.

**Comment by jcechace** [comment]
@nilshartmann  awesome job on this one, thank you for contributing!

**Comment by nilshartmann** [comment]
@jcechace one more question 😔: what is still the purpose of `build-all.sh` and esp. `image-builds-auth.yml`? 

- If one would run `image-builds-auth.yml` (or `build-all`) with `PUSH_IMAGES`set to `true`, I think that would override/remove the arm images in the docker hub/quay registry, that might have been pushed before with the `*-multiplatform.sh` scripts. 
- Without setting `PUSH_IMAGES` to `true` I see no differences to the "new" scripts, esp. no reason, why `image-builds-auth.yml` logs into docker...

(Just out of curiosity: I also wonder where the arm64 image were built that are currently in [dockerhub](https://hub.docker.com/r/debezium/connect-base/tags), as your PR https://github.com/debezium/debezium/pull/3950 hasn't been merged yet?)

**Pull request merged**
